content/blog/en/dojo-middleware.md

```diff
@@ -0,0 +1,213 @@
+
+> src/widgets/PhoneValidator.tsx
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+import phoneNumberMiddleware from '../middleware/phoneNumberMiddleware';
+
+import * as css from './styles/PhoneValidator.m.css';
+
+interface Properties {
+	phone: string;
+}
+
+const factory = create({ phoneNumberMiddleware }).properties<Properties>();
+
+export const PhoneValidator = factory(function PhoneValidator({ children, middleware: { phoneNumberMiddleware } }) {
+	const { valid, value } = phoneNumberMiddleware();
+	let validCss = '';
+	if (value.length) {
+		validCss = valid ? css.valid : css.invalid;
+	}
+	return <div classes={[css.root, validCss]}>{children()}</div>;
+});
+
+export default PhoneValidator;
+```
+
+The `PhoneValidator` uses middleware that returns the a `valid` property that is either `true` or `false`. The middle will also return the `value` of the phone number that was tested. Based on whether the phone number is valid or not, it will provide CSS for a red or green border.
+
+Notice that the `phone` property to the middleware is not passed explicitly to the middleware. By providing the `phoneNumberMiddleware` as a middleware to the `PhoneValidator` widget, the middleware gets access to the properties of the widget. Let's see what that looks like:
+
+> src/middleware/phoneNumberMiddleware.tsx
+
+```tsx
+import { create } from '@dojo/framework/core/vdom';
+
+const factory = create().properties<{ phone?: string }>();
+
+export const phoneNumberMiddleware = factory(({ properties }) => {
+	return () => {
+		// extract the `phone` value from the properties of
+		// the parent widget
+		const { phone } = properties();
+		// test the phone number
+		const valid = /^\(?(\d{3})\)?[- ]?(\d{3})[- ]?(\d{4})$/.test(phone || '');
+		return {
+			valid,
+			value: phone
+		};
+	};
+});
+
+export default phoneNumberMiddleware;
+```
+
+The middleware returns a function that tests the phone number and returns whether it is valid or not.
+
+View this middleware example in a sample phone validation application.
+
+!(https://codesandbox.io/embed/dojo-custom-middleware-w1f7m?fontsize=14&module=%2Fsrc%2Fmiddleware%2FphoneNumberMiddleware.ts)
+
+## Geolocation Middleware
+
+You could also create some fun middleware that interacts with the DOM of your widgets. For example, there are the [`intersection`](https://dojo.io/learn/middleware/available-middleware#intersection) and [`resize`](https://dojo.io/learn/middleware/available-middleware#resize) middleware.
+
+You could use a similar pattern to grab the browser's geolocation coordinates.
+
+> src/middleware/geolocation.ts
+
+```ts
+import { create } from '@dojo/framework/core/vdom';
+import icache from '@dojo/framework/core/middleware/icache';
+
+const factory = create({ icache });
+
+type Coords = Pick<Coordinates, 'latitude' | 'longitude'>;
```

Comment:
Where does the `Coordinates` type come from? In the interest of keeping this simple type wise would it not be better to just type latitude and longitude?

Reply:
`Coordinates` comes from TypeScripts `lib.dom.d.ts`